### DISCUSSION TITLE
Inkonsistente Daten bei der Wallbox (mqtt)

### DISCUSSION THREAD

**Discussion by frizzle0815** [solectrus/solectrus#3510] [General]
Hallo Georg, 

ich beobachte seit einiger Zeit das Problem, dass der Verbrauch durch meine Wallbox in Solectrus nicht richtig aufgezeichnet wird. Es handelt sich um eine openWB, die ihre Daten über mqtt zur Verfügung stellt. Die Daten werden mit QoS 0 gesendet, was ich glaube ich auch nicht ändern kann.. vielleicht ist das Teil des Problems.

Ich habe mehrere Sachverhalte:

1. Der in Solectrus angezeigte Verbrauch ist generell zu hoch. Heute morgen in der Zeit von 2:30 Uhr bis 4:00 Uhr wurden laut Wallbox z.B. 6,05 kWh geladen, aber Solectrus zeigt 9,5 kWh an. Eine enorme Differenz. Meine Vermutung geht dahin, dass Solectrus insbesondere nicht mitbekommt, wenn der Ladevorgang beendet wird. Ich habe im Diagramm dann häufiger (aber auch nicht immer) eine flache Linie auf Höhe des letzten Verbrauchs. Damit rechnet Solectrus dann vermutlich irgendwie weiter und erzeugt so den zu hohen Verbrauch.

2. In Solectrus stimmt in diesen Fällen die Tages-Summe nicht mit Woche/Monat überein. Für Tag wird jetzt gerade 9,524 kWh angezeigt, im Wochen- und Monatsdiagramm sind es 9,329 kWh. Hier ist Woche/Monat niedriger. Ich habe aber auch Fälle, wo bei Tag 31,829 kWh und bei Woche/Monat für diesen Tag 38,464 kWh angezeigt wird. Laut Wallbox waren es an dem Tag tatsächlich nur 22,09 kWh...

3. Der Power Splitter ist hierbei auch merkwürdig. Für das Laden heute Nacht wird mir angezeigt, dass 51% aus PV geladen wären. Das kann niemals stimmen. Ein geringer Anteil stammte aus dem Hausakku, was ja als PV angezeigt werden würde. Das waren aber niemals 51%, vielleicht eher so 20% bis der Hausakku leer war und dann nur noch aus dem Netz geladen wurde. Ich vermute mal, dass der zu hohe Verbrauch (9,329/9,524 gegenüber tatsächlichen 6,05 kWh) bei PV mit aufgeschlagen wurde... das wäre stimmig.

Mir stellen sich mehrere Fragen:

1. Wenn Solectrus nicht mitbekommt, dass der Ladevorgang beendet wurde, müsste doch aber zumindest sichergestellt sein, dass nicht unendlich mit dem falschen Wert weiter gerechnet wird. Die Wallbox scheint den 0-Verbrauch nur einmal zu senden (QoS 0) und wenn Solectrus das verpasst (woran könnte das liegen?!), dann sollte Solectrus irgendwann selbst von einem 0-Verbrauch ausgehen, wenn für ein Intervall x keine neuen Daten über mqtt ankommen.

2. Warum die Abweichungen zwischen Tag und Woche/Monat? Dafür habe ich gar keine Erklärung... zu Punkt 1 könnte man noch sagen, dass der Endanwender irgendwie ausreichende Datenqualität sicherstellen muss. Aber das hier sieht nach einem Problem in Solectrus aus.

3. Ein Verbesserungsvorschlag: Ich empfinde es grundsätzlich als ineffizient, dass Solectrus versuchen muss aus dem aktuellen Ertrag/Verbrauch auf den Gesamtwert zu schließen. Da die Daten maximal alle 5-10 Sekunden aktualisiert werden, wird das doch immer zu Ungenauigkeiten führen. Die scheinen grundsätzlich nur marginal zu sein, aber mein Problem oben zeigt, dass es auch anders sein kann. Oder was ist, wenn der mqtt-Server mal zeitweise ausfällt?
Nun ist es zumindest bei mir so, dass Solectrus gar nicht den Ertrag/Verbrauch rechnerisch ermitteln müsste, weil eigentlich für alles exakte Messwerte zur Verfügung gestellt werden könnten. Ich kann aus dem Wechselrichter abrufen, wie hoch der Ertrag des Tages war, wieviel der Akku geladen/entladen wurde, wie hoch der Hausverbrauch war. Bei mir lese ich sogar den Hauszähler aus, so dass ich über die Zählerstände weiß, wie hoch der Netzbezug bzw. die Einspeisung war. Und zu guter Letzt kann ich auch bei der Wallbox exakte Zählerwerte auslesen.
Ließen sich diese Daten nicht integrieren?

**Comment by ledermann** [comment]
Alle Berechnungen von SOLECTRUS basieren darauf, dass kontinuierlich Messwerte hereinkommen. Eine Wallbox muss also auch dann Messwerte liefern, wenn der Stromverbrauch gerade 0 ist. Der Grund ist, dass SOLECTRUS aus Performancegründen zur Berechnung der Energie nicht die genauere Integral-Methode von InfluxDB verwendet, sondern die deutlich schnellere Methode der Stunden-Durchschnitts-Summe. Dies geht aber schief, wenn die Messwerte Lücken enthalten. Nur QoS auf 1 zu bringen, genügt also **nicht**. Kannst du dafür sorgen, dass die Wallbox auch 0-Werte kontinuierlich sendet? Dann wäre alles gut.

Da der Power-Splitter auf präzise (und kontinuierlich) gemessenen Verbräuchen basiert, sind Abweichungen hier nur Folgefehler.
Ähnlich sieht es bei den Diagrammen für Woche/Monat/Jahr aus.

Zu deinem Verbesserungsvorschlag: Ja, SOLECTRUS berechnet selbst aus der Leistung (in Watt) die Energie (in kWh). Das liegt darin begründet, dass viele Geräte die Energie gar nicht liefern, daher möchte ich das auch so belassen. Die Abweichungen, die durch das 5-Sekunden-Intervall entstehen, sind der Erfahrung nach verschwindend gering.

Das Grundproblem ist also, dass SOLECTRUS nicht gut damit klarkommt, wenn Messwerte Lücken enthalten. Hier überlege ich mir schon länger eine Lösung, um die präzise Integral-Funktion zu benutzen (die mit Lücken klarkommt), ohne deren Performance-Nachteile in Kauf zu nehmen. Die Idee sind vorberechnete Summen per Background-Job. Das ist aber eine sehr große und elementare Änderung, die viel Zeit zur Implementierung benötigen wird, damit ist kurzfristig leider nicht zu rechnen.

**Reply by frizzle0815** [reply]
Ich frage die Wallbox jetzt mit meinem Gira Homeserver ab und leite die Werte von dort alle 10 Sekunden an ein neues Topic im mqtt Server weiter, das ich dann für Solectrus benutze. Damit sollten jetzt kontinuierlich Daten zur Verfügung stehen.

**Reply by ledermann** [reply]
Ja, das hört sich gut an. Vielleicht kannst du Bescheid geben, ob der erwünschte Effekt auch eintritt.

**Reply by frizzle0815** [reply]
Nach der Umstellung zeigt mir die Tagesansicht heute 4,07 kWh, die heute Nacht geladen wurden. Das ist exakt der Wert, den auch meine Wallbox angibt, also perfekt!
In der Wochen-/Monatsansicht wird aber leider 4,913 kWh angezeigt.

**Reply by frizzle0815** [reply]
Diese Abweichungen tauchen übrigens nicht nur bei der Wallbox auf. Bei den übrigen Werten sind die Differenzen aber kleiner.

Ich vergleiche immer die Säulen im Wochen-/Monatsdiagramm mit den Werten oben in der Tagesanzeige.

Was mir jetzt aber auch aufgefallen ist: Wenn ich die Säulen in der Wochenansicht aufsummiere, stimmt dies nicht mit der darüber angezeigten Summe überein.

**Reply by ledermann** [reply]
Ok, die Abweichung bei den Diagrammen für größere Zeiträume ist ein bekanntes Problem, siehe https://github.com/solectrus/solectrus/issues/2032. Es hat nichts mit MQTT zu tun, sondern tritt auf, wenn das Diagramm einen nicht-abgeschlossenen Zeitraum betrifft. Das Problem wird verschwinden, wenn der Zeitraum abgeschlossen ist, also vollständig in der Vergangenheit liegt.

Diese ärgerliche, aber temporäre Ungenauigkeit werde ich noch angehen.

**Reply by frizzle0815** [reply]
Super, danke 😊
Dann ist das Thema für mich ab hier erledigt 👍
